lib/src/kbasesearchengine/main/DecorateWithNarrativeInfo.java

```diff
@@ -0,0 +1,121 @@
+        this.searchInterface = searchInterface;
+        this.wsClient = wsClient;
+    }
+
+    public SearchTypesOutput searchTypes(SearchTypesInput params, String user)
+            throws Exception {
+        return searchInterface.searchTypes(params, user);
+    }
+
+    public Map<String, TypeDescriptor> listTypes(String uniqueType)
+            throws Exception {
+        return searchInterface.listTypes(uniqueType);
+    }
+
+    public SearchObjectsOutput searchObjects(SearchObjectsInput params, String user)
+            throws Exception {
+        SearchObjectsOutput searchObjsOutput = searchInterface.searchObjects(params, user);
+        searchObjsOutput.setAccessGroupNarrativeInfo(addNarrativeInfo(searchObjsOutput.getObjects(),
+                searchObjsOutput.getAccessGroupNarrativeInfo()));
+        return searchObjsOutput;
+    }
+
+    public GetObjectsOutput getObjects(GetObjectsInput params, String user)
+            throws Exception {
+        GetObjectsOutput getObjsOutput = searchInterface.getObjects(params, user);
+        getObjsOutput.setAccessGroupNarrativeInfo(addNarrativeInfo(getObjsOutput.getObjects(),
+                getObjsOutput.getAccessGroupNarrativeInfo()));
+        return getObjsOutput;
+    }
+
+    private Map<Long, Tuple5 <String, Long, String, String, String>> addNarrativeInfo(
+            List<ObjectData> objects,
+            Map<Long, Tuple5 <String, Long, String, String, String>> accessGroupNarrInfo)
+                throws IOException, JsonClientException {
+
+        long wsId;
+        Tuple9 <Long, String, String, String, Long, String, String,
+                String,Map<String,String>> wsInfo;
+        Tuple5 <String, Long, String, String, String> tempNarrInfo =
+                new Tuple5<>();
+        final Map<Long, Tuple5 <String, Long, String, String, String>> retVal = new HashMap<>();
+
+        if (accessGroupNarrInfo != null) {
+            retVal.putAll(accessGroupNarrInfo);
+        }
+
+        List<Long> wsIdsList = new ArrayList<Long>();
+
+        for (ObjectData objData: objects) {
+
+            GUID guid = new GUID(objData.getGuid());
+
+            String storageCode = guid.getStorageCode();
+
+            if (STORAGE_CODE.equals(storageCode)) {
+                wsId = guid.getAccessGroupId();
+                if (!wsIdsList.contains(wsId))
+                    wsIdsList.add(wsId);
+            }
+            else {
+                // TODO: handle other storage types?
+            }
+        }
+        for (long workspaceId: wsIdsList) {
+            if (!retVal.containsKey(workspaceId)) {
+                // get workspace info meta data
+                WorkspaceEventHandler weh = new WorkspaceEventHandler(
+                                                new CloneableWorkspaceClientImpl(wsClient));
+                wsInfo = weh.getWorkspaceInfo(workspaceId);
+                tempNarrInfo.setE3(wsInfo.getE4());
```

Comment:
This is the timestamp right? If so parse to epoch millis. Using a text timestamp was a huge mistake.

Reply:
Done. Changed the type of this field to 'long' in the tuple.